randovania/server/bot.py

```diff
@@ -53,16 +56,652 @@ async def reply_for_preset(message: discord.Message, versioned_preset: Versioned
+                    await async_race_admin_msg("*%s forfeit their run of '%s'.*" % (player_name, race_name))
+                else:
+                    raise Exception
+            elif command[0] == "!vod":
+                if state != AsyncRacerState.SUBMITTED_TIME:
+                    await message.author.send("*Please submit your time first with the `!finish hh:mm:ss` commands.*")
+                    return
+                else:
+                    vod_url = message.content.split(" ")[1] # this is case sensitive because youtube links are case sensitive
+                    if async_races[race_idx].players[player_idx].time == DNF:
+                        async_races[race_idx].players[player_idx].state = AsyncRacerState.DNF
+                    else:
+                        async_races[race_idx].players[player_idx].state = AsyncRacerState.SUBMITTED_VOD
+                    async_races[race_idx].players[player_idx].vod_url = vod_url
+                    async_races[race_idx].players[player_idx].vod_timestamp = time.time()
+                    diff =  (async_races[race_idx].players[player_idx].vod_timestamp - async_races[race_idx].players[player_idx].time_timestamp) - ASYNC_RACE_SUBMIT_VOD_TIME_LIMIT_M*60
+                    await message.author.send("*Your run has been recorded. Please wait for results to be announced in #async-racing-results before discussing the seed or making your VOD public.*")
+                    if diff > 0:
+                        await message.author.send("**Warning: You have submitted your VOD %d minutes late. While this does not automatically disqualify your run, the race organizer has been notified of this infraction.**" % (diff/60))
+                        await async_race_admin_msg("*%s, %s was %d minutes late submitting their VOD for '%s'.*" % (async_race_admin_ping(), player_name, (diff/60), race_name))
+                    else: 
+                        await async_race_admin_msg("*%s submitted the vod of their run for '%s':*\n`%s`" % (player_name, race_name, vod_url))
+                
+                is_race_done = True
+                for player in async_races[race_idx].players:
+                    if player.vod_url == "":
+                        is_race_done = False
+                        break
+                if is_race_done:
+                    await async_race_admin_msg("*%s, All racers in '%s' have completed their run. Close submissions and announce results with `!asyncrace end %s`, or announce results and start a new race with the same players using `!asyncrace turnover %s <permalink>`*" % (async_race_admin_ping(), race_name, race_name, race_name))
+            else:
+                raise Exception
+        except Exception as e:
+            exc_type, exc_obj, exc_tb = sys.exc_info()
+            fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
+            print(exc_type, fname, exc_tb.tb_lineno)
```

Comment:
Use the logging module, so this goes to nicer places. And to be honest, makes a lot of this code redundant if you just do a `logging.exception(f"Except when handling DM: {e}")`